python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +91,453 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+    a: typing.Union[list[list[fem.Form]], fem.Form],
+    L: typing.Union[list[fem.Form], fem.Form],
+    P: typing.Union[list[list[fem.Form]], list[fem.Form], fem.Form, None],
+    assembly_type: fem.AssemblyType,
+) -> tuple[PETSc.Mat, PETSc.Vec, PETSc.Vec, PETSc.Mat | None]:  # type: ignore
+    """Create data-structures used in PETSc NEST solvers
+
+    Args:
+        a: The compiled bi-linear form(s)
+        L: The compiled linear form(s)
+        P: The compiled preconditioner form(s)
+        assembly_type: The type of NEST solver to use
+    Returns:
+        PETSc datastructures for the matrix A, vectors x and b, and preconditioner P
+    """
+
+    matrix_creator: typing.Union[None, typing.Callable[[PETSc.Mat], typing.Any]] = None  # type: ignore
+    vector_creator: typing.Union[None, typing.Callable[[PETSc.Vec], typing.Any]] = None  # type: ignore
+    if assembly_type == fem.AssemblyType.default:
+        matrix_creator = create_matrix
+        vector_creator = create_vector
+    elif assembly_type == fem.AssemblyType.block:
+        matrix_creator = create_matrix_block
+        vector_creator = create_vector_block
+    elif assembly_type == fem.AssemblyType.nest:
+        matrix_creator = create_matrix_nest
+        vector_creator = create_vector_nest
+    else:
+        raise ValueError("Unsupported SNES type")
+    A = matrix_creator(a)  # type: ignore
+    b = vector_creator(L)  # type: ignore
+    x = vector_creator(L)  # type: ignore
+    P = None if P is None else matrix_creator(P)  # type: ignore
+    return A, x, b, P
+
+
+class SNESSolver:
+    def __init__(
+        self,
+        F: typing.Union[dolfinx.fem.Form, ufl.form.Form],
+        u: dolfinx.fem.Function,
+        J: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        P: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        assembly_type: fem.AssemblyType = fem.AssemblyType.default,
+        bcs: typing.Optional[list[dolfinx.fem.DirichletBC]] = None,
+        form_compiler_options: typing.Optional[dict] = None,
+        jit_options: typing.Optional[dict] = None,
+    ):
+        """Initialize a PETSc-SNES solver."""
+        self._u = u
+        self._snes, self._x = create_snes_solver(
+            F, self._u, J, P, assembly_type, bcs, form_compiler_options, jit_options
+        )
+        if assembly_type == fem.AssemblyType.default:
+            self._copy_solution = copy_solution
+            self._replace_solution = replace_solution
+        elif assembly_type == fem.AssemblyType.block:
+            self._copy_solution = copy_solution_block
+            self._replace_solution = replace_solution_block
+        elif assembly_type == fem.AssemblyType.nest:
+            self._copy_solution = copy_solution_nest
+            self._replace_solution = replace_solution_nest
+        else:
+            raise ValueError("Unsupported Assembly type")
+
+    def set_options(self, options: dict):
```

Comment:
Remove this and pass up the options at construction time? This is consistent with e.g. LinearProblem.